kinta-lib/src/main/kotlin/com/dailymotion/kinta/integration/appgallery/AppGalleryIntegration.kt

```diff
@@ -0,0 +1,444 @@
+        }
+        val token = KintaEnv.get(KintaEnv.Var.APPGALLERY_TOKEN) ?: acquireToken(clientId, clientSecret)
+
+        val data = JsonObject(mapOf())
+
+        val request = getRequest(clientId, token)
+                .url(HttpUrl.parse("$API_URL/publish/v2/app-submit")!!.newBuilder()
+                        .addQueryParameter("appId", appId)
+                        .addQueryParameter("releaseType", "1")
+                        .build()
+                )
+                .post(RequestBody.create(JSON, data.toString()))
+                .build()
+
+        val response = OkHttpClient.Builder().build().newCall(request).execute()
+        check(response.isSuccessful) {
+            "Error releasing app to AppGallery : ${response.body()?.string()}"
+        }
+        val jsonResult = Json.nonstrict.parseJson(response.body()?.string().orEmpty()).jsonObject
+        check(jsonResult["ret"]?.jsonObject?.get("code")?.intOrNull ?: -1 == 0){
+            "Error releasing app to AppGallery : $jsonResult"
+        }
+    }
+
+    /**
+     * Upload a listing
+     */
+    fun uploadListing(
+            client_id: String? = null,
+            client_secret: String? = null,
+            package_name: String? = null,
+            listing: GooglePlayIntegration.ListingResource
+    ) {
+        val clientId = client_id ?: KintaEnv.getOrFail(KintaEnv.Var.APPGALLERY_CLIENT_ID)
+        val clientSecret = client_secret ?: KintaEnv.getOrFail(KintaEnv.Var.APPGALLERY_CLIENT_SECRET)
+        val packageName = package_name ?: KintaEnv.getOrFail(KintaEnv.Var.APPGALLERY_PACKAGE_NAME)
+        val appId = try {
+            getAppId(clientId, clientSecret, packageName)
+        } catch (e: Exception) {
+            /** We probably have an invalid token **/
+            acquireToken(clientId, clientSecret)
+            getAppId(clientId, clientSecret, packageName)
+        }
+        val token = KintaEnv.get(KintaEnv.Var.APPGALLERY_TOKEN) ?: acquireToken(clientId, clientSecret)
+
+        val data = JsonObject(
+                mapOf(
+                        "lang" to JsonPrimitive(listing.language),
+                        "appName" to JsonPrimitive(listing.title),
+                        "appDesc" to JsonPrimitive(listing.description),
+                        "briefInfo" to JsonPrimitive(listing.shortDescription)
+                )
+        )
+
+        val request = getRequest(clientId, token)
+                .url(HttpUrl.parse("$API_URL/publish/v2/app-language-info")!!.newBuilder()
+                        .addQueryParameter("appId", appId)
+                        .build()
+                )
+                .put(RequestBody.create(JSON, data.toString()))
+                .build()
+
+        Logger.i("Uploading listing ${listing.language} to AppGallery...")
+        val response = OkHttpClient.Builder().build().newCall(request).execute()
+        check(response.isSuccessful) {
+            "Error uploading listing : ${response.body()?.string()}"
```

Comment:
I think this leaks a network connection and `response.close` should be called: https://square.github.io/okhttp/4.x/okhttp/okhttp3/-response-body/

Reply:
Yes, you're right!